packages/new-editor/src/components/editor/variables/master/Variables.tsx

```diff
@@ -49,14 +50,100 @@ export const Variables = ({ variables, onSelection }: VariablesProps) => {
     }
   });
 
+  const getSelectedRow = () => {
+    return table.getRowModel().rowsById[Object.keys(selection.tableState.rowSelection!)[0]];
+  };
+
+  const getParentIndexes = (row: Row<Variable>) => {
+    const parentId = row.parentId;
+    if (!parentId) {
+      return [];
+    }
+    return parentId.split('.').map(index => Number(index));
+  };
+
+  const deepCopyVariables = () => {
+    return deepCopyVariablesRecursive(variables);
+  };
+
+  const deepCopyVariablesRecursive = (variables: Variable[]) => {
+    return variables.map(variable => {
+      const newVariable = { ...variable };
+      newVariable.children = deepCopyVariablesRecursive(variable.children);
+      return newVariable;
+    });
+  };
+
+  const getVariable = (variables: Variable[], indexes: number[]): Variable | undefined => {
+    if (indexes.length === 0) {
+      return;
+    }
+    return getVariableRecursive(variables, [...indexes]);
+  };
+
+  const getVariableRecursive = (variables: Variable[], indexes: number[]): Variable => {
+    const variable = variables[indexes.shift()!];
+    if (indexes.length === 0) {
+      return variable;
+    }
+    return getVariableRecursive(variable.children, indexes);
+  };
+
+  const selectVariable = (indexes: number[], variable?: Variable) => {
+    setSelectedVariable(variable);
+    if (indexes.length === 0) {
+      selection.options.onRowSelectionChange({});
+    } else {
+      selection.options.onRowSelectionChange({ [`${indexes.join('.')}`]: true });
+    }
+  };
+
+  const adjustSelectionBeforeDeletion = (index: number, children: Variable[], parentIndexes: number[], parent?: Variable) => {
+    switch (children.length) {
+      // variable is the last remaining child of its parent -> select parent
+      case 1:
+        selectVariable(parentIndexes, parent);
+        break;
+      // variable is the last child in the list of children of its parent -> select previous variable
+      case index + 1:
+        selectVariable([...parentIndexes, index - 1], children[index - 1]);
+        break;
+      // select next variable
+      default:
+        setSelectedVariable(children[index + 1]);
+        break;
+    }
+  };
+
+  const addVariable = () => {}; // TODO: Implementation
+
+  const deleteVariable = () => {
```

Comment:
I think this whole variable data modify code can easily be tested with some unit tests. E.g. the like form editor data.ts it is doing.